### PR TITLE
feat(nntp): split transfer and metadata connection budgets

### PR DESCRIPTION
**Pull request by jackster1209**
## Summary

Separates provider-wide connection capacity from article-transfer concurrency. Users can now define a **Provider Connection Limit** (total NNTP connections for the account) independently from **Transfer Connections** (the hard cap for `BODY`/`ARTICLE` work). The remaining capacity becomes available for lightweight metadata operations (`STAT`, `HEAD`, `DATE`, health checks, connectivity probes) with bounded burst access to idle transfer slots.

**Before:** `MaxConnections` controlled both the provider ceiling and the transfer concurrency. Auto-tune recommended the smallest connection count near the throughput knee, which could leave paid provider connections entirely unused.

**After:** A provider configured with 50 connections and 20 transfer connections keeps article transfers at the benchmarked sweet spot while making 30–40 connections available for metadata work — without ever exceeding the provider ceiling.

## Motivation

Users whose provider accounts allow significantly more connections than the throughput-optimal transfer count were leaving paid capacity on the table. Metadata-heavy workloads (Sonarr/Radarr imports, health checks, provider availability probes) competed for the same small connection budget as transfers. This feature lets InfiniDysk use more of what users already pay for without sacrificing the transfer optimization Auto-tune was designed to provide.

## Design

- **Single physical pool.** No second connection pool is created. A new `ProviderConnectionAdmission` layer sits in front of the existing `ConnectionPool` and classifies each NNTP operation as Transfer or Metadata before granting a lease.
- **Hard transfer cap.** `BODY`, `ARTICLE`, and their pipelined variants are classified as Transfer and never exceed the configured `MaxTransferConnections`.
- **Bounded metadata burst.** Metadata receives a guaranteed base allocation (`P - T`) plus up to `floor(T / 2)` borrowed from idle transfer capacity. Metadata can never consume the entire transfer allocation.
- **Transfer-biased reclaim with starvation protection.** When transfer work arrives and metadata has borrowed capacity, in-flight metadata operations finish normally. Released slots normally return to waiting transfers first, while one metadata waiter may run after at most eight consecutive transfer grants so health and control work cannot starve.
- **Existing priority scheduling preserved.** Streaming Priority (high/low odds) continues to operate within the admission layer.
- **Null = legacy admission mode.** `MaxTransferConnections` is nullable. Existing providers with no value configured retain the original shared transfer/metadata admission scheduling. Merely opening or saving an existing provider does not activate split budgeting; the safer Auto-tune ceiling and busy-pool keep-alive behavior apply in both modes.

### Runtime invariants

```text
ActiveTransfers + ActiveMetadata  ≤  EffectiveProviderConnectionLimit
ActiveTransfers                   ≤  EffectiveTransferConnectionLimit
ActiveMetadata                    ≤  BaseMetadataCapacity + floor(T / 2)
```

### Budget math

```text
P = effective Provider Connection Limit
T = min(configured MaxTransferConnections, P)

Base metadata    = P - T
Burst allowance  = floor(T / 2)
Metadata max     = base + burst
```

A one-connection provider still gets at least one opportunistic metadata lease for recovery probes while its transfer slot is idle.

### Backend (C#)

- **`UsenetProviderConfig`** — added nullable `MaxTransferConnections` property.
- **`ConfigManager`** — validation: transfer connections must be ≥ 1 and ≤ `MaxConnections`.
- **`ProviderConnectionBudget`** — pure record struct with a static `Calculate()` method for testable budget math.
- **`ProviderConnectionAdmission`** — operation-aware admission gate with priority-aware wait queues, bounded metadata burst, and transfer-priority reclaim. Disposable with clean shutdown of waiters.
- **`ConnectionLock` / acquisition lifecycle** — operation gates are collected in one idempotent composite lease, attached through a single callback, and released in reverse acquisition order. Any exception after physical acquisition disposes the physical lock and all unattached leases.
- **`MultiConnectionNntpClient`** — acquires operation leases before the physical pool lock when `MaxTransferConnections` is configured; classifies operations via `ClassifyConnectionKind()` and exposes allocation-free routing state.
- **`UsenetBenchmarkService`** — Auto-tune recommendation now applies to Transfer Connections, never rewrites or probes above Provider Connection Limit, and reports prominently when throughput is still climbing at that ceiling. Pipelining-only runs receive a separately validated transfer test count.
- **`PrometheusMetrics`** — extended `nzbdav_nntp_pool_connections` with `transfer_active`, `metadata_active`, `transfer_waiting`, `metadata_waiting` states; extended `nzbdav_nntp_pool_max_connections` with `transfer_configured`, `transfer_effective`, `metadata_base`, `metadata_burst`, `metadata_max` limits.
- **`ProviderLatencyTracker`** — updated for split scheduling awareness.
- **`SupportPackService`** — support packs now include admission snapshot data.
- **`GetProviderUsageController`** — API response includes live connection budget breakdown when available.

### Frontend (TypeScript/React)

- **Provider editor modal** — new Transfer Connections field with validation. Metadata Capacity preview computed from configured values before save, then from live runtime data after.
- **Provider cards** — display Transfer Connections and Metadata Capacity range when split mode is active; show "Legacy shared-pool connection scheduling" otherwise.
- **Auto-tune apply** — recommendation targets Transfer Connections; Provider Connection Limit is never changed.
- **`provider-connection-budget.ts`** — client-side budget calculation and formatting utilities with tests.
- **`provider-autotune.ts`** — transfer recommendation application logic with tests.

### Docs

- **`usenet.md`** — new Connection Budgets section with the three-field model, budget math, examples table, and legacy-mode callout.
- **`headless.md`** — updated example JSON with `MaxTransferConnections`, added explanation.
- **`first-run.md`** — updated provider setup table for new field names.
- **`prometheus.md`** — documented new admission metrics and label values.

### Tests and benchmark coverage

| Test file | Coverage |
|-----------|----------|
| `ProviderConnectionBudgetTests` | Budget math edge cases (equal limits, single connection, learned limit smaller than configured) |
| `ProviderConnectionAdmissionTests` | Admission, transfer cap, metadata burst, burst exhaustion, reclaim priority, cancellation, disposal, lock-free routing state |
| `ConnectionLockTests` | One composite callback, reverse lease release, idempotent cleanup |
| `ConnectionPoolWarmConnectionTests` | Bounded busy-pool skip, idle-only borrowing, distinct probes, 100/0 contention recovery |
| `MultiConnectionConnectionBudgetTests` | End-to-end admission through `MultiConnectionNntpClient` for transfer vs. metadata operations |
| `MultiConnectionStatsPipelinedTests` | Extended pipelined STAT paths under split scheduling |
| `ProviderConnectionSnapshotTests` | Snapshot includes admission data when present |
| `UsenetProvidersValidationTests` | Config validation for transfer connections (null, valid, below 1, above max) |
| `GetProviderUsageConnectionBudgetTests` | API response shape with and without split mode |
| `DetectKneeTests` | Transfer recommendation capping and still-climbing ceiling guidance |
| `BenchmarkUsenetConnectionRequestTests` | Invalid ceilings/counts, transfer-over-provider rejection, pipelining-only fallback, verification bounds |
| `ProviderLatencyTrackerTests` | Latency tracking with budget-aware providers |
| `PrometheusMetricsTests` | Admission metric labels set and removed correctly |
| `provider-connection-budget.test.ts` | Frontend budget calculation and formatting |
| `provider-autotune.test.ts` | Frontend auto-tune transfer recommendation and benchmark request validation |
| `ProviderSelectionBenchmarks` | Multi-provider selection under concurrent admission acquire/release |

## Backward compatibility and behavior changes

- Providers without `MaxTransferConnections` retain legacy shared-pool admission scheduling; no split admission layer is created.
- Existing persisted configs are not modified. No migration step is required.
- Auto-tune no longer probes above Provider Connection Limit, including for legacy providers. Raise that limit first when a ceiling result is still climbing.
- Keep-alive probes now time out and skip when admission or the physical pool is busy instead of waiting without bound or opening a replacement socket.
- Pooled, BackupAndStats, and BackupOnly provider types all work with the new budgeting.

## Integration with PR #1208

PR #1208 must rebase after this branch lands. Its combined acquisition path must acquire health gate → provider admission → physical pool, add both operation leases to the shared composite owner, and attach exactly one cleanup callback. The rebased PR must add a health-context STAT regression covering success, cancellation, pool-acquisition failure, and cleanup attachment failure before merge.

## How to test

1. **Legacy provider (no `MaxTransferConnections`):** Confirm transfer/metadata admission scheduling remains shared-pool, while Auto-tune respects Provider Connection Limit and busy keep-alives skip promptly.
2. **Split provider (e.g. 50/20):** Verify transfer concurrency never exceeds 20, metadata can use up to 40, total never exceeds 50.
3. **Auto-tune:** Run a benchmark, apply recommendation — Transfer Connections should be set, Provider Connection Limit should remain unchanged.
4. **Prometheus:** Scrape `/metrics` and confirm `transfer_active`, `metadata_active`, and limit labels appear for split providers.
5. **Concurrent workload:** Run a large Sonarr import alongside Plex playback; metadata should utilize the additional capacity while transfers stay at their optimized level.

## Summary by CodeRabbit

* **New Features**
  * Added separate provider-wide and transfer connection limits for Usenet providers.
  * Added operation-aware budgeting with metadata capacity, burst allowances, and live activity counts.
  * Added budget details to provider usage, diagnostics, and monitoring metrics.
  * Added UI support for transfer limits, capacity previews, and transfer-only Auto-tune recommendations.

* **Bug Fixes**
  * Benchmark recommendations now stay within configured provider limits.
  * Improved connection cleanup during disposal, failures, and cancellation.

* **Documentation**
  * Updated setup, configuration, deployment, monitoring, and Auto-tune guidance.

### CODE REVIEW

File: backend/Services/Benchmark/UsenetBenchmarkService.cs [hunk 1]
```diff
@@ -750,7 +804,11 @@ private static List<int> BuildLevels(int configuredMaxConnections, BenchmarkProf
             if (prev.MegaBytesPerSec > 0 && (peak.MegaBytesPerSec - prev.MegaBytesPerSec) / prev.MegaBytesPerSec > 0.08)
             {
                 stillClimbing = true;
-                warnings.Add("Speed was still climbing at the highest level tested — a faster line or even more connections may help.");
+                warnings.Add(
+                    configuredProviderLimit is { } limit && peak.Connections >= limit
+                        ? $"Speed was still climbing at your Provider Connection Limit ({limit}). " +
+                          "Raise that limit and re-run Auto-tune to test higher transfer counts."
+                        : "Speed was still climbing at the highest level tested — a faster line or even more connections may help.");
```
[the remark this answers is not in the document]
Comment (hoivikaj) [reply backend/Services/Benchmark/UsenetBenchmarkService.cs]: Fixed in afc73b8e. Still-climbing guidance now distinguishes the 50-connection benchmark ceiling from a higher configured Provider Connection Limit, with regression coverage for 51/50.

**Comment by hoivikaj** [comment]
@coderabbitai review
